File: backend/app/Models/JobApplication.php
<?php

namespace App\Models;

use Database\Factories\JobApplicationFactory;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasOne;
use Illuminate\Database\Eloquent\SoftDeletes;

#[Fillable([
    'job_vacancy_id',
    'user_id',
    'alumni_id',
    'status',
    'cover_letter',
    'cv_data',
    'cv_path',
    'portfolio_path',
    'applied_at',
])]
class JobApplication extends Model
{
    /** @use HasFactory<JobApplicationFactory> */
    use HasFactory, HasUuids, SoftDeletes;

    public const STATUSES = [
        'submitted',
        'reviewing',
        'shortlisted',
        'interview',
        'accepted',
        'rejected',
        'withdrawn',
    ];

    /**
     * Get the attributes that should be cast.
     *
     * @return array<string, string>
     */
    protected function casts(): array
    {
        return [
            'status' => 'string',
            'cv_data' => 'array',
            'applied_at' => 'datetime',
        ];
    }

    /**
     * The vacancy this application belongs to.
     */
    public function vacancy(): BelongsTo
    {
        return $this->belongsTo(JobVacancy::class, 'job_vacancy_id');
    }

    /**
     * The user (alumni account) that applied.
     */
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    /**
     * The linked alumni profile (when present).
     */
    public function alumni(): BelongsTo
    {
        return $this->belongsTo(Alumni::class);
    }

    /**
     * The hiring result recorded when the application is accepted.
     */
    public function acceptance(): HasOne
    {
        return $this->hasOne(JobAcceptance::class);
    }
}

File: mobile/lib/models/job_application.dart
/// Lamaran kerja alumni (job application).
class JobApplication {
  final String id;
  final String jobVacancyId;
  final String status; // submitted|reviewing|shortlisted|interview|accepted|rejected|withdrawn
  final String? coverLetter;
  final String? appliedAt;
  final JobApplicationVacancy? vacancy;

  const JobApplication({
    required this.id,
    required this.jobVacancyId,
    required this.status,
    this.coverLetter,
    this.appliedAt,
    this.vacancy,
  });

  factory JobApplication.fromJson(Map<String, dynamic> json) {
    return JobApplication(
      id: json['id'] as String? ?? '',
      jobVacancyId: json['job_vacancy_id'] as String? ?? '',
      status: json['status'] as String? ?? 'submitted',
      coverLetter: json['cover_letter'] as String?,
      appliedAt: json['applied_at'] as String?,
      vacancy: json['vacancy'] is Map<String, dynamic>
          ? JobApplicationVacancy.fromJson(json['vacancy'] as Map<String, dynamic>)
          : null,
    );
  }
}

class JobApplicationVacancy {
  final String id;
  final String title;
  final String companyName;
  final String? employmentType;
  final String? location;

  const JobApplicationVacancy({
    required this.id,
    required this.title,
    required this.companyName,
    this.employmentType,
    this.location,
  });

  factory JobApplicationVacancy.fromJson(Map<String, dynamic> json) {
    return JobApplicationVacancy(
      id: json['id'] as String? ?? '',
      title: json['title'] as String? ?? '',
      companyName: json['company_name'] as String? ?? '',
      employmentType: json['employment_type'] as String?,
      location: json['location'] as String?,
    );
  }
}
